packages/dashboard/src/components/composed/Debugger/Sources/Source/LineNumber.tsx

```diff
@@ -0,0 +1,95 @@
+import {
+  highlightedTextTag,
+  closingHighlightedTextTag
+} from "src/components/composed/Debugger/utils";
+import { createStyles } from "@mantine/core";
+import { useDash } from "src/hooks";
+
+interface LineNumberProps {
+  lineHasHighlighting: boolean;
+  lineNumber: number;
+  lineNumberGutterWidth: number;
+  sourceId: string;
+}
+
+const useStyles = createStyles(() => ({
+  lineNumber: {
+    "&:hover": {
+      cursor: "pointer"
+    },
+    "display": "flex"
+  },
+  spacer: {
+    minHeight: 22,
+    minWidth: 22,
+    width: 22,
+    marginRight: 5
+  },
+  breakpointSpacerContainer: {
+    minHeight: 22,
+    minWidth: 22,
+    width: 22,
+    display: "flex",
+    marginRight: 5,
+    alignItems: "center"
+  },
+  breakpointSpacer: {
+    width: 16,
+    height: 16,
+    backgroundColor: "#FA5252",
+    borderRadius: 25
+  }
+}));
+
+function LineNumber({
+  lineHasHighlighting,
+  lineNumberGutterWidth,
+  sourceId,
+  lineNumber
+}: LineNumberProps): JSX.Element {
+  const { classes } = useStyles();
+  const {
+    state: {
+      debugger: { breakpoints }
+    },
+    operations: { toggleDebuggerBreakpoint }
+  } = useDash()!;
+  // if the line contains highlighting we highlight the line number as well
+  const lineNumberDisplay = lineHasHighlighting
+    ? `<span class="${classes.lineNumber}">` +
+      "&nbsp;".repeat(lineNumberGutterWidth - lineNumber.toString().length) +
+      highlightedTextTag +
+      lineNumber +
+      closingHighlightedTextTag +
+      "&nbsp;&nbsp;" +
+      `</span>`
+    : `<span class="${classes.lineNumber}">` +
+      "&nbsp;".repeat(lineNumberGutterWidth - lineNumber.toString().length) +
+      lineNumber +
+      "&nbsp;&nbsp;" +
+      `</span>`;
```

Comment:
Can we move this to the part where the component returns? So we can avoid `dangerouslySetInnerHTML`.

Reply:
I don't think we can since we need to generate raw html. Notice that we have a static class to designate text as highlighted. We generate the html for the source separately from this and then highlight the number in this step. I suppose we could just use a separate class...but we also need to be able to set whitespace as well for source formatting.

Reply:
There are two things here:

The first thing is related to my comment [here](https://github.com/trufflesuite/truffle/pull/6058#discussion_r1254978658) about not putting debugger-related styles in `<Global />`, so the class that does the highlight belongs here.

The second thing is that even if the class name _has_ to be static, we still don't need to write markup in string and then `dangerouslySetInnerHtml`. It would just be

```tsx
return (
  ...
  <div> { /* Don't dangerouslySetInnerHtml */ }
    <span class={`${classes.lineNumber} ${highlighted ? "your-static-class" : undefined}`}>
      {lineNumber}
    </span>
  </div>
  ...
)
```
But again, replace `"your-static-class"` with a new `classes.` class that you define here, not globally.

Reply:
I made a comment on the comment on `MantineGlobal` around line 54. Currently the ".truffle-debugger-text-highlighted" class gets applied in about 3 different places and so it is a global. In my reply I ask if we should consider architecting this differently. The static class doesn't apply to everyting in `lineNumber`, only a portion so I don't think we can do exactly what you suggest unless I'm being dense and not fully understanding right now.

Reply:
updated!